[lit] add --max-retries-per-test execution option

kwk opened the pull request:
When packaging LLVM we've seen arbitrary tests fail.
It happened sporadically and most of the times the test
do work if they are run a second time on the next day.

The tests themselves were always different and we didn't
know ahead of time which ones we wanted to re-run.
That's we filter-out a lot of `libomp` and `libarcher` tests [1].

This change allows us to set
`LIT_OPTS="--max-retries-per-test=12"`
when running any "check-XXX" build target. Then any lit test
will at most be re-run 12 times, unless there's an `ALLOW_RETRIES:`
in one of the test scripts that's specifying a different value
than `12`. `12` is just an example here, any positive integer
will work.

Please note, that this only adds the possibility to re-run
lit tests. It does not actually do it until the caller specifies
`--max-retries-per-test=<POSITIVE_INT>` either on a call to `lit` or
in `LIT_OPTS`.

Also note, that one can still use `ALLOW_RETRIES:` in test scripts
and it will always rule over `--max-retries-per-test`. When
`--max-retries-per-test` is set too low, but the `config.test_retry_attempts`
is high enough, it works as well.

Any option in the list below overrules its predecessor:

* `--max-retries-per-test`
* `config.test_retry_attempts`
* `ALLOW_RETRIES` keyword

From the above options to re-run tests, `--max-retries-per-test` is the only one that doesn't require a change in the test scripts or the test config.

[1]: https://src.fedoraproject.org/rpms/llvm/blob/rawhide/f/llvm.spec#_2326

Downstream PR to make use of the `--max-retries-per-test` option: https://src.fedoraproject.org/rpms/llvm/pull-request/434
Downstream ticket: https://issues.redhat.com/browse/LLVM-145

nikic reviewed (commented):
I don't think that this should be an openmp specific option. Lit should provide a generic option to set the (default) number of retry attempts, which is available for all test suites.

kwk commented:
> I don't think that this should be an openmp specific option. Lit should provide a generic option to set the (default) number of retry attempts, which is available for all test suites.

Fair enough. Let me make that change then.

kwk committed:
[lit] add --max-retries-per-test execution option

When packaging LLVM we've seen arbitrary tests fail.
It happened sporadically and most of the times the test
do work if they are run a second time on the next day.

The tests themselves were always different and we didn't
know ahead of time which ones we wanted to re-run.
That's we filter-out a lot of `libomp` and `libarcher` tests [1].

This change allows us to set
`LIT_OPTS="--max-retries-per-test=12"`
when running any "check-XXX" build target. Then any lit test
will at most be re-run 12 times, unless there's an `ALLOW_RETRIES:`
in one of the test scripts that's specifying a different value
than `12`. `12` is just an example here, any positive integer
will work.

Please note, that this only adds the possibility to re-run
lit tests. It does not actually do it until the caller specifies
`--max-retries-per-test=<POSITIVE_INT>` either on a call to `lit` or
in `LIT_OPTS`.

Also note, that one can still use `ALLOW_RETRIES:` in test scripts
and it will always rule over `--max-retries-per-test`. When
`--max-retries-per-test` is set too low, but the `config.test_retry_attempts`
is high enough, it works as well.

Any option in the list below overrules its predecessor:

* `--max-retries-per-test`
* `config.test_retry_attempts`
* `ALLOW_RETRIES` keyword

[1]: https://src.fedoraproject.org/rpms/llvm/blob/rawhide/f/llvm.spec#_2326

Downstream PR to make use of the `--max-retries-per-test` option: https://src.fedoraproject.org/rpms/llvm/pull-request/434
Downstream ticket: https://issues.redhat.com/browse/LLVM-145

nikic reviewed (commented):
Can you please also document the option in https://github.com/llvm/llvm-project/blob/main/llvm/docs/CommandGuide/lit.rst?

kwk committed:
Fix test suite name

kwk committed:
Adjust help text of --max-retries-per-test

kwk commented on llvm/utils/lit/tests/Inputs/max-retries-per-test/allow-retries-no-test_retry_attempts/test.py:
```diff
@@ -0,0 +1,23 @@
+# ALLOW_RETRIES: 3
+# RUN: "%python" "%s" "%counter"
```
In this call: `RUN: "%python" "%s" "%counter"`, `%counter` is the `sys.argv[1]`. Or am I completely misunderstanding this?

RoboTux replied:
No sorry it's me, I read it as %s %t %counter. By the way why not use simply %t and add .counter to it?

kwk commented on llvm/utils/lit/tests/Inputs/max-retries-per-test/allow-retries-test_retry_attempts/test.py:
```diff
@@ -0,0 +1,23 @@
+# ALLOW_RETRIES: 3
+# RUN: "%python" "%s" "%counter"
```
It would maybe be possible to share the test file across two of the tests that don't modify `ALLOW_RETRIES:` in the test file. But the test file is being used in several other places as well. And so I continued in that habit.

RoboTux replied:
Sure, I was thinking of 2 test files: one with ALLOW_RETRIES and one without.

kwk commented on llvm/utils/lit/tests/Inputs/max-retries-per-test/no-allow-retries-test_retry_attempts/test.py:
```diff
@@ -0,0 +1,22 @@
+# RUN: "%python" "%s" "%counter"
```
I don't think this is a good idea. See answer above.

kwk committed:
Update command guide

kwk commented on llvm/utils/lit/lit/cl_arguments.py:
```diff
@@ -199,6 +199,13 @@ def parse_args():
+        "--max-retries-per-test",
+        dest="maxRetriesPerTest",
+        metavar="N",
+        help="Maximum number of allowed retry attempts per test (NOTE: ALLOWED_RETRIES keyword always takes precedence)",
```
Done in fa4abfd4ac40a3c6c97c6d4162b772099a781fd0

RoboTux reviewed (commented):
LGTM for what it's worth.
